Members/csv upload

oscarwang20 opened the pull request:
### Summary <!-- Required -->

This pr implements a feature in the admin edit users page to upload a csv file containing a user's email and any other updated information that can be updated through the edit users page. The csv file must have the user's email as a field, and the field names must be in `camelCase` (`"first name"` -> `"firstName"`)

### Notion/Figma Link <!-- Optional -->

https://www.notion.so/cornelldti/Import-and-Update-Member-Info-By-CSV-62243888a1484a22bde912ca167fd31f?pvs=4

### Test Plan <!-- Required -->

`members.csv`
<img width="508" alt="image" src="https://github.com/cornell-dti/idol/assets/69322572/de5aaa46-bc4b-4a6c-b79c-587c7a6c8550">
File selected:
<img width="305" alt="image" src="https://github.com/cornell-dti/idol/assets/69322572/1b34a201-6d0f-4765-b78c-3738d4f2d95b">
File uploaded:
<img width="309" alt="image" src="https://github.com/cornell-dti/idol/assets/69322572/1590241d-4c47-400b-a1f4-6aaaaf48f4a1">
Changes:
<img width="1003" alt="image" src="https://github.com/cornell-dti/idol/assets/69322572/f262fe04-c2f2-4ebb-9b42-2f02abbc4e7a">
Changes (cont):
<img width="991" alt="image" src="https://github.com/cornell-dti/idol/assets/69322572/77e4d986-2dba-4f98-8733-5a78fb1d0b44">

### Notes <!-- Optional -->

<!--- List any important or subtle points, future considerations, or other items of note. -->

### Breaking Changes <!-- Optional -->

<!-- Keep items that apply: -->

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +128,75 @@ export default function AddUser(): JSX.Element {
+          const currMember = allMembers.find((mem) => mem.netid === getNetIDFromEmail(m.email));
+          if (currMember) {
+            members.push({
+              netid: getNetIDFromEmail(m.email),
```
Since you already calculated `getNetIDFromEmail(m.email)` in `currMember`, and you do it for every member, and you use it again here, how about you store that value in a variable first so we don't have to repeatedly calculate it?

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -182,6 +254,27 @@ export default function AddUser(): JSX.Element {
               </div>
             </Card.Content>
+            <Card.Content>
+              <div className={`ui one buttons ${styles.fullWidth}`}>
```
UX improvement ideas: how about we only show the upload button if a file has been chosen already? And let's add some text here that says "Add/update users through CSV". And maybe let's even include a template file that the admin can download?

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +128,75 @@ export default function AddUser(): JSX.Element {
+              about: m.about || '',
+              subteams: [m.subteam] || [],
+              formerSubteams: m.formerSubteams || [],
+              role: m.role || ('' as Role),
```
Getting the role right is important. We should add an error if they don't include a "role" column. And I think we should also throw an error if not all the roles specified are one of the roles available "lead, tpm, pm, developer, designer, business". See if you can add this requirement without code duplication (i.e. redefining the role list again).

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +128,75 @@ export default function AddUser(): JSX.Element {
+              github: m.github || '',
+              hometown: m.hometown || '',
+              about: m.about || '',
+              subteams: [m.subteam] || [],
```
Can we do a validation check to make sure that the subteam actually exists?

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +128,75 @@ export default function AddUser(): JSX.Element {
+              hometown: m.hometown || '',
+              about: m.about || '',
+              subteams: [m.subteam] || [],
+              formerSubteams: m.formerSubteams || [],
```
`formerSubteams` should have the same validation check as the `subteam` one.

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +128,75 @@ export default function AddUser(): JSX.Element {
+              subteams: [m.subteam] || currMember.subteams,
+              formerSubteams: m.formerSubteams || currMember.formerSubteams,
+              role: m.role || currMember.role,
+              roleDescription: getRoleDescriptionFromRoleID(m.role)
```
Can you do a null check on `m.role` here too, like you did below?

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +128,75 @@ export default function AddUser(): JSX.Element {
+        json.forEach((m) => {
+          const currMember = allMembers.find((mem) => mem.netid === getNetIDFromEmail(m.email));
+          if (currMember) {
+            members.push({
```
I like to avoid code duplication if possible, but not while sacrificing readability. But I think we can do a little better here. Thoughts on this?

Create a generic "EMPTY_MEMBER" object and place it up top as a constant. Line 84 in this file can use this object to set the state.

```
const EMPTY_MEMBER = {
        firstName: '',
        lastName: '',
        pronouns: '',
        email: '',
        role: '' as Role,
        graduation: '',
        major: '',
        doubleMajor: '',
        minor: '',
        website: '',
        linkedin: '',
        github: '',
        hometown: '',
        about: '',
        subteams: [],
        formerSubteams: []
}
```

Then we use this same object to loop through the keys of the object:

```
const member = {};
Object.keys(EMPTY_MEMBER).forEach((field) => { 
 member[field] = m[field] ?? currMember[field]
}
members.push(member)
```
And then down below we can do 

```
const member = {...EMPTY_MEMBER};
Object.keys(EMPTY_MEMBER).forEach((field) => { 
 if(m[field])
   member[field] = m[field]
}
members.push(member)
```

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -182,6 +254,27 @@ export default function AddUser(): JSX.Element {
+                accept=".csv"
+                onChange={(e) => setCsvFile(e.target.files?.[0])}
+              />
+              {uploadStatus ? <p>{uploadStatus}</p> : undefined}
```
UX suggestion: What if we made this red, so it shows up easier?

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -74,6 +75,8 @@ export default function AddUser(): JSX.Element {
     isCreatingUser: false
   });
+  const [csvFile, setCsvFile] = useState<File | undefined>(undefined);
+  const [uploadStatus, setUploadStatus] = useState<string>('');
```
More like, `uploadError`, at least based on how it's being used.

oscarwang20 replied:
It's also used to display a success message indicating how many members were uploaded, so `uploadError` is a little imprecise.
<img width="300" alt="image" src="https://github.com/cornell-dti/idol/assets/69322572/44b99de6-1b54-4f45-b506-17249acaf5d5">

I've changed the hook to include a status and a message which is rendered as either a green success message or a red error message.
<img width="303" alt="image" src="https://github.com/cornell-dti/idol/assets/69322572/0682d5dd-eb14-4cac-a9a2-8686cbaf25d7">
<img width="303" alt="image" src="https://github.com/cornell-dti/idol/assets/69322572/7f4f10f7-7ac2-4f95-bb77-f911daef6888">

andrew032011 replied:
I agree. I didn't see that you had the `Uploaded` message originally. I like what you have now.

andrew032011 reviewed (changes requested):
Very nice work! This looks great. I think the leads are going to love this when the newbies get here.

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +146,46 @@ export default function AddUser(): JSX.Element {
+      const json = await csvtojson().fromString(csv);
+      const allNetIds = allMembers.map((m) => m.netid);
+      const members: IdolMember[] = [];
+      if (json[0].email && json[0].role) {
```
The requirement for `role` should only be enforced for new members. For existing members, we should just keep the role the same if it's unspecified.

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +146,46 @@ export default function AddUser(): JSX.Element {
+            members.push(member);
+          }
+        });
+        members.forEach((m) => {
```
Hm, can we upload each member one by one by calling `MembersAPI.updateMember`/`setMember` in the `json.forEach` function? Instead of first populating a list, and then going through the list again to call the API for each one?

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +146,46 @@ export default function AddUser(): JSX.Element {
+            Object.keys(EMPTY_MEMBER).forEach((field) => {
+              if (m[field]) member[field] = m[field];
+            });
+            members.push(member);
```
I think these member objects are missing the `netid` and `roleDescription` fields.

andrew032011 replied:
Yep, those two fields are missing.

<img width="965" alt="Screenshot 2023-09-19 at 12 33 00 PM" src="https://github.com/cornell-dti/idol/assets/59291082/fc3ffab0-8ba5-4aed-a08d-3dfef2de19f5">

andrew032011 replied:
You can insert these fields right before you push the `member` to `members` array (or if you make the revision I said above, just add the two fields right before you call `MembersAPI`.

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,13 +146,64 @@ export default function AddUser(): JSX.Element {
     });
   }
 
+  function RenderUploadStatus(): JSX.Element {
```
nit: `UploadStatus`, since it is just a component. No need for `Render`

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,13 +146,64 @@ export default function AddUser(): JSX.Element {
 
+  function RenderUploadStatus(): JSX.Element {
+    if (uploadStatus) {
+      switch (uploadStatus.status) {
```
nit: How about just 
```
<p className={uploadStatus.status === 'error' ? styles.errorMessage : styles.successMessage}>{`${uploadStatus.msg}`}</p>;
```
and no switch statement?

andrew032011 commented:
UI bug? I wouldn't block releasing this change from this but just something to note

<img width="328" alt="Screenshot 2023-09-19 at 12 08 55 PM" src="https://github.com/cornell-dti/idol/assets/59291082/f5f03a78-1463-4225-ae7f-79f1237c61ea">

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +129,75 @@ export default function AddUser(): JSX.Element {
+          github: m.github || currMember.github,
+          hometown: m.hometown || currMember.hometown,
+          about: m.about || currMember.about,
+          subteams: [m.subteam] || currMember.subteams,
```
Empty arrays are truthy in JavaScript! If you test this out with a csv and you don't specify the subteam, the existing subteams will actually be deleted. Do `m.subteam ? [m.subteam] : currMember.subteams`

andrew032011 replied:
<img width="961" alt="Screenshot 2023-09-19 at 10 20 56 PM" src="https://github.com/cornell-dti/idol/assets/59291082/54851579-9c91-442e-a47a-b85841020e4f">

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +129,75 @@ export default function AddUser(): JSX.Element {
+          github: m.github || '',
+          hometown: m.hometown || '',
+          about: m.about || '',
+          subteams: [m.subteam] || [],
```
See above. rewrite as `m.subteam ? [m.subteam] : []`

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +129,75 @@ export default function AddUser(): JSX.Element {
+    if (csvFile) {
+      const csv = await csvFile.text();
+      const json = await csvtojson().fromString(csv);
+      if (json[0].email) {
```
This logic is a little confusing to follow and would make refactoring/adding on difficult. I'd rewrite it like this:
```
if(!json[0].email) {
  setUploadStatus({ status: 'error', msg: 'Error: No email field!' });
}
else if(!json[0].role) {
  setUploadStatus({ status: 'error', msg: 'Error: No role field!' });
}
else {
  processJson(json);
  setUploadStatus({ status: 'success', msg: `Uploaded ${json.length} members` });
}
```

andrew032011 reviewed (changes requested):
How about some validation checks for role and subteam? I'd ignore an entry entirely if the role/subteam was wrong. And I'd list that out in my error message all the entries that were wrong.

oscarwang20 commented:
Instead of checking for missing fields using `json[0]`, `uploadUsersCsv` now filters out the invalid fields. I'll try to make the error messages more descriptive so it's easier to identify the invalid rows.

I had to change it because the state updates are asynchronous and aren't always caught as `uploadUsersCsv` runs.

Alternatively, I could change the filtering to filter out the well-formatted members and submit those while displaying error messages for the improperly formatted users. Wondering what would be less annoying to deal with...

andrew032011 commented:
> Alternatively, I could change the filtering to filter out the well-formatted members and submit those while displaying error messages for the improperly formatted users. Wondering what would be less annoying to deal with...

I vote for failing the entire process, but tell us specifically rows are wrong/off. I think it could get more annoying if you had some of them uploaded and some of them not because it can get hard to keep track.

andrew032011 commented:
If you're missing the email or role column altogether though, that should be an overall error that you raise, not specific to any row (unless I guess they left that box empty).

andrew032011 commented on frontend/src/components/Admin/AddUser/AddUser.tsx:
```diff
@@ -125,6 +135,84 @@ export default function AddUser(): JSX.Element {
+          !m.email ||
+          !m.role ||
+          !ALL_ROLES.includes(m.role as Role) ||
+          !validSubteams.includes(m.subteam)
```
You need to check if we've even specified a subteam in the first place (i.e. `m.subteam && !validSubteams.includes(m.subteam)`), since the `subteam` field is optional.

oscarwang20 commented:
- Added two new error messages for missing email and role fields
- Error messages now list each invalid row and their invalid fields

Felt might also be worthwhile to check if a member's subteam is in both `subteam` and `formerSubteams`. Excel and Google Sheets also format cells with comma separated values as strings, so I changed the way that's handled as well.

andrew032011 commented:
> Also, can we include a template .csv here? One that contains all the columns? At the moment, there's no intuitive way someone would know our field names unless they looked at the code.

And some helper text about what's required would be nice. Similar to what I'm adding for Candidate Decider here: https://github.com/cornell-dti/idol/pull/496

andrew032011 reviewed (changes requested):
Very nice work, and excellent error handling! The messages are very helpful. A few nits.

I manually tested your changes using the following CSVs. Please add an example CSV that includes ALL the possible columns and add some helper text that explains the format we expect the input in, and that `email` and `role` are required columns.

[Missing_one_email_role_misspell_subteam_Member_CSV_Upload_Example.csv](https://github.com/cornell-dti/idol/files/12711261/Missing_one_email_role_misspell_subteam_Member_CSV_Upload_Example.csv)
[Missing_email_Member_CSV_Upload_Example.csv](https://github.com/cornell-dti/idol/files/12711262/Missing_email_Member_CSV_Upload_Example.csv)
[Member_CSV_Upload_Example.csv](https://github.com/cornell-dti/idol/files/12711263/Member_CSV_Upload_Example.csv)
